Fix spokes ordering and add tests for priority uniqueness

jkonecny12 opened the pull request:
~The "KernelWarningSpoke" and "UnsupportedHardwareSpoke" seems to be switched. Not sure if this is happening only sometimes or all the time but sorting the spokes in the test should solve this for good.~

~Probably introduced by: 0aa6f1507a8b4bacebca7dcccf7bda4b873e6fba~

~**EDIT:**~
~If two standalone spokes have the same priority then we depending on implementation of sorted algorithm which can differ between architectures.~

**EDIT2:**
See https://github.com/rhinstaller/anaconda/pull/2435#issuecomment-614684320 for explanation.

This already happened in the past see:
https://bugzilla.redhat.com/show_bug.cgi?id=929177

jkonecny12 commented:
jenkins, test this please

jkonecny12 commented:
**UPDATED:**
* Or better changed the implementation. Remove the old solution and change the code with a adding new test.

poncovka commented on tests/nosetests/pyanaconda_tests/simple_ui_test.py:
```diff
@@ -156,3 +197,30 @@ def gui_test(self, window_group, window, meh, builder, proxy_getter):
+        """
+        # Create the interface.
+        from pyanaconda.ui.gui import GraphicalUserInterface
+        self.interface = GraphicalUserInterface(self.storage, self.payload)
```
Please, extend the existing tests. Don't copy-paste them.

jkonecny12 replied:
I'm seeing this as another test. The first test is testing if the spokes are correctly assigned to hubs and they are there. This new test is testing if the spokes are explicitly positioned.
It feels like overloading test cases.

poncovka commented on pyanaconda/ui/common.py:
```diff
@@ -480,6 +480,8 @@ class StandaloneSpoke(Spoke):
     preForHub = None
     postForHub = None
 
+    priority = 100
```
It looks like the default priority should be zero.

jkonecny12 replied:
This is default priority from the past but we can change it to make it stable with the current state.

poncovka commented on pyanaconda/ui/gui/spokes/network.py:
```diff
@@ -1578,7 +1578,6 @@ class NetworkStandaloneSpoke(StandaloneSpoke):
     uiFile = "spokes/network.glade"
 
     preForHub = SummaryHub
-    priority = 10
```
Please, don't change the priorities when it is not necessary. There might be addons that rely on these values.

jkonecny12 replied:
OK

poncovka commented on tests/nosetests/pyanaconda_tests/simple_ui_test.py:
```diff
@@ -56,6 +56,10 @@ def hubs(self):
 
+    @property
+    def ordered_action_classes(self):
+        return self.interface._orderActionClasses(self.action_classes, self.hubs)
```
Why don't we just fix the method `_orderActionClasses`?

jkonecny12 replied:
See my answer below in https://github.com/rhinstaller/anaconda/pull/2435#issuecomment-614055745.

poncovka commented:
From https://docs.python.org/3/library/functions.html#sorted:

> The built-in sorted() function is guaranteed to be stable. A sort is stable if it guarantees not to change the relative order of elements that compare equal — this is helpful for sorting in multiple passes (for example, sort by department, then by salary grade).

The problem is with the initial list of spokes. We need to sort these spokes by name first.

jkonecny12 commented:
It's totally my fault that I did not looked on the official documentation to verify my discoveries from a past. I guess this could be a difference between the old Python 2 code and current Python 3. Good idea to fix spokes ordering instead of this solution.
However, there is a question if we don't want to require explicit ordering than using the sort. For me personally it seems that the explicit ordering is the better option and much easier to spot what is happening in the code.

pep8speaks commented:
Hello @jkonecny12! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pyanaconda/ui/__init__.py`](https://github.com/rhinstaller/anaconda/blob/90c16a99eae085a2893a5824d3d2988e8b785d19/pyanaconda/ui/__init__.py):

> [Line 94:100](https://github.com/rhinstaller/anaconda/blob/90c16a99eae085a2893a5824d3d2988e8b785d19/pyanaconda/ui/__init__.py#L94): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)
> [Line 157:97](https://github.com/rhinstaller/anaconda/blob/90c16a99eae085a2893a5824d3d2988e8b785d19/pyanaconda/ui/__init__.py#L157): [E502](https://duckduckgo.com/?q=pep8%20E502) the backslash is redundant between brackets
> [Line 158:17](https://github.com/rhinstaller/anaconda/blob/90c16a99eae085a2893a5824d3d2988e8b785d19/pyanaconda/ui/__init__.py#L158): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 183:100](https://github.com/rhinstaller/anaconda/blob/90c16a99eae085a2893a5824d3d2988e8b785d19/pyanaconda/ui/__init__.py#L183): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)
> [Line 189:100](https://github.com/rhinstaller/anaconda/blob/90c16a99eae085a2893a5824d3d2988e8b785d19/pyanaconda/ui/__init__.py#L189): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 99 characters)











* In the file [`tests/nosetests/pyanaconda_tests/simple_ui_test.py`](https://github.com/rhinstaller/anaconda/blob/90c16a99eae085a2893a5824d3d2988e8b785d19/tests/nosetests/pyanaconda_tests/simple_ui_test.py):

> [Line 133:9](https://github.com/rhinstaller/anaconda/blob/90c16a99eae085a2893a5824d3d2988e8b785d19/tests/nosetests/pyanaconda_tests/simple_ui_test.py#L133): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 204:39](https://github.com/rhinstaller/anaconda/blob/90c16a99eae085a2893a5824d3d2988e8b785d19/tests/nosetests/pyanaconda_tests/simple_ui_test.py#L204): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 207:39](https://github.com/rhinstaller/anaconda/blob/90c16a99eae085a2893a5824d3d2988e8b785d19/tests/nosetests/pyanaconda_tests/simple_ui_test.py#L207): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 210:39](https://github.com/rhinstaller/anaconda/blob/90c16a99eae085a2893a5824d3d2988e8b785d19/tests/nosetests/pyanaconda_tests/simple_ui_test.py#L210): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 217:9](https://github.com/rhinstaller/anaconda/blob/90c16a99eae085a2893a5824d3d2988e8b785d19/tests/nosetests/pyanaconda_tests/simple_ui_test.py#L217): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)

jkonecny12 commented:
**UPDATED:**
* Implemented notes from @poncovka.
* Added ordering of spokes before dividing them to the action classes in `_orderActionClasses()`.
* Added test for this ordering to prevent happening this in the future.
* Change the default priority to `0`.
* Move priority test to the existing test functions to remove those giant patching a preparation.
* Make ``ui/__init__.py`` almost pep8 compliant (without changing API).

I'm solving this on two fronts.
1. Adding priority to solve the Unsupported HW and Kernel spokes ordering and adding test to require setting different priority if there are two spokes with the same priority on the same hub.
2. Sorting spokes by name before sorting them by priority.

The first one will solve the second for us but not for spokes. This way we will make sure that we are not depending just on priority/names and addons can make use of that.

poncovka commented on pyanaconda/ui/__init__.py:
```diff
@@ -147,13 +153,21 @@ def _collectActionClasses(self, module_pattern_w_path, standalone_class):
+            standalones.extend(
+                collect(module_pattern,
+                        path,
+                        lambda obj: issubclass(obj, standalone_class) and
```
Sometimes it is better to define a new function (it can be nested) instead of of using lambda functions.

jkonecny12 replied:
I've only reformatted the code not change it. But yeah why not.

poncovka commented on pyanaconda/ui/__init__.py:
```diff
@@ -163,13 +177,22 @@ def _orderActionClasses(self, spokes, hubs):
-        return actionClasses
+            action_classes.extend(
+                sorted(
+                    filter(lambda obj, h=hub: getattr(obj, "preForHub", None) == h,
```
I don't think this is readable. I would replace the lambda function with a function. It looks like it could be used also in tests. Same below.

jkonecny12 replied:
Fixed.

poncovka commented on tests/nosetests/pyanaconda_tests/simple_ui_test.py:
```diff
@@ -70,6 +75,20 @@ def _get_category_names(self, hub_class):
+                self.assertNotIn(
+                    priority,
+                    res,
+                    "Spokes {} and {} have the same priority!".format(res[priority], name))
```
I would create the message first and then call the assert method, so all arguments can be on the same line.

jkonecny12 replied:
Fixed.

poncovka commented on tests/nosetests/pyanaconda_tests/simple_ui_test.py:
```diff
@@ -110,6 +129,16 @@ def tui_test(self, proxy_getter):
+        spokes = self.ordered_action_classes
+
+        for hub in self.hubs:
+            pre_spokes = filter(lambda obj, h=hub: getattr(obj, "preForHub", None) == h, spokes)
```
I don't understand why we define `h=hub` in these lambda functions.

jkonecny12 replied:
I've just copied this. I guess it's some old convention?

jkonecny12 replied:
Fixed.

jkonecny12 commented on tests/nosetests/pyanaconda_tests/simple_ui_test.py:
```diff
@@ -156,3 +185,36 @@ def gui_test(self, window_group, window, meh, builder, proxy_getter):
+            pre_spokes = filter(lambda obj, h=hub: getattr(obj, "preForHub", None) == h, spokes)
+            self._check_spokes_priority_uniqueness(pre_spokes)
+            post_spokes = filter(lambda obj, h=hub: getattr(obj, "postForHub", None) == h, spokes)
+            self._check_spokes_priority_uniqueness(post_spokes)
```
Fixed.

jkonecny12 commented on tests/nosetests/pyanaconda_tests/simple_ui_test.py:
```diff
@@ -156,3 +185,36 @@ def gui_test(self, window_group, window, meh, builder, proxy_getter):
+        list2 = [SpokeC, SpokeB, SpokeA]
+
+        for spoke in list1:
+            spoke.preForHub = hub
```
Fixed.

jkonecny12 commented on tests/nosetests/pyanaconda_tests/simple_ui_test.py:
```diff
@@ -156,3 +185,36 @@ def gui_test(self, window_group, window, meh, builder, proxy_getter):
+
+        # the input list ordering shouldn't matter sorting should be based on class names
+        self.assertEqual(UserInterface._orderActionClasses(list1, [hub]),
+                         UserInterface._orderActionClasses(list2, [hub]))
```
Fixed.

jkonecny12 commented:
**UPDATED:**
* Implemented suggestions from Vendy.

jkonecny12 commented on pyanaconda/ui/__init__.py:
```diff
@@ -163,13 +180,28 @@ def _orderActionClasses(self, spokes, hubs):
+            action_classes.append(hub)
+            action_classes.extend(
+                sorted(
+                    UserInterface._filter_spokes_by_post_for_hub_reference(ordered_spokes, hub),
```
This is a static method so I don't have self.
